hazelcast/src/main/java/com/hazelcast/map/impl/MapEntryReplacingEntryProcessor.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl;
+
+import com.hazelcast.map.EntryProcessor;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+public class MapEntryReplacingEntryProcessor<K, V> implements EntryProcessor<K, V, V>, IdentifiedDataSerializable {
+
+    BiFunction<? super K, ? super V, ? extends V> function;
+
+    public MapEntryReplacingEntryProcessor() {
+    }
+
+    public MapEntryReplacingEntryProcessor(BiFunction<? super K, ? super V, ? extends V> function) {
+        this.function = function;
+    }
+
+    @Override
+    public V process(Map.Entry<K, V> entry) {
+        V newValue = function.apply(entry.getKey(), entry.getValue());
+        entry.setValue(newValue);
+        return null;
+    }
+
+    @Override
+    public int getFactoryId() {
+        return MapDataSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return MapDataSerializerHook.MAP_ENTRY_REPLACING_PROCESSOR;
+    }
+
+    @Override
+    public void writeData(ObjectDataOutput out) throws IOException {
+        out.writeObject(function);
+    }
+
+    @Override
+    public void readData(ObjectDataInput in) throws IOException {
+        function = in.readObject();
+    }
+
+    @Nullable
+    @Override
+    public EntryProcessor<K, V, V> getBackupProcessor() {
+        return null;
```

Comment:
I think we should simply not override this method to execute the same on the backup replicas.

Reply:
@mmedenjak - I guess I should do the same for `KeyValueConsumingEntryProcessor` (used in `forEach()` implementation) - can you please confirm?

https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java#L1123

Reply:
I've checked the others. Since we don't set the new value, means we're not mutating the entry, which means it's read-only.
Now that I think about it, it may be the case that some users use `IMap.forEach` to mutate the value. We have two options then:
• we continue supporting only read-only use of `forEach`. This way, users mutating the values may get surprised but they can always switch to using regular EPs. Also, this way we don't introduce a massive overhead for read-only uses of `forEach`, meaning we don't create defensive copies, serialize after the lambda has been invoked, invoke entry listeners, WAN replication, etc. for every single value
• we defensively act as if the value was mutated. That means both read-only and mutating lambdas will work correctly.

Between the two, I'd go with option 1. It's a bit surprising but users can always simply switch to a regular EP to mutate the values and we avoid the prohibitively expensive overhead of defensively copying, serializing, notifying listeners, etc. for read-only uses of `forEach`.

In other words, you can leave that one as-is.

Reply:
Agree with you. `forEach()` is a bit subjective one!

Reply:
I guess there's also the marker interface of `ReadOnly` which can be added to the lambda but... I'm not sure if everyone is aware of it and I'd rather keep this for another PR if we decide to change the current behaviour.